### ISSUE TITLE
Rule-change (`no-return-assign`) - allow assignment in side effect position of comma operator

### ISSUE THREAD

**Issue by EECOLOR** [eslint/eslint#9418]
Update: @platinumazure suggested I filled out the change-rule template:

**What rule do you want to change?**
`no-return-assign`

**Does this change cause the rule to produce more or fewer warnings?**
Fewer

**How will the change be implemented? (New option, new default behavior, etc.)?**
Change current default behavior (`except-parens`)

**Please provide some example code that this change will affect:**

```js
1. return (x = y, x)
2. return ((x = y), x)
3. return (x = y)
4. return x = y
5. return (x, x = y)
6. return (x, (x = y))
```

**What does the rule currently do for this code?**
1. Reports error
2. Does not report error
3. Does not report error
4. Reports error
5. Reports error
6. Does not report error

**What will the rule do after it's changed?**

Only 1. will change, 1. will no longer report an error because the assignment is in a position of a sequence (comma operator) that is designed for side effects ([see MDN example](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Comma_Operator#Processing_and_then_returning)).

---
**Original issue:** `no-return-assign` - false positive in combination with comma operator

The assignment in the code below is not in 'return' position:
```js
[].map(result => (result['x'] = 'y', result))
```

[Example in the online demo](https://eslint.org/demo/#eyJ0ZXh0IjoiW10ubWFwKHJlc3VsdCA9PiAocmVzdWx0Wyd4J10gPSAneScsIHJlc3VsdCkpXG4iLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOjYsInNvdXJjZVR5cGUiOiJzY3JpcHQiLCJlY21hRmVhdHVyZXMiOnt9fSwicnVsZXMiOnsiY29uc3RydWN0b3Itc3VwZXIiOjIsIm5vLWNhc2UtZGVjbGFyYXRpb25zIjoyLCJuby1jbGFzcy1hc3NpZ24iOjIsIm5vLWNvbXBhcmUtbmVnLXplcm8iOjIsIm5vLWNvbmQtYXNzaWduIjoyLCJuby1jb25zb2xlIjoyLCJuby1jb25zdC1hc3NpZ24iOjIsIm5vLWNvbnN0YW50LWNvbmRpdGlvbiI6Miwibm8tY29udHJvbC1yZWdleCI6Miwibm8tZGVidWdnZXIiOjIsIm5vLWRlbGV0ZS12YXIiOjIsIm5vLWR1cGUtYXJncyI6Miwibm8tZHVwZS1jbGFzcy1tZW1iZXJzIjoyLCJuby1kdXBlLWtleXMiOjIsIm5vLWR1cGxpY2F0ZS1jYXNlIjoyLCJuby1lbXB0eS1jaGFyYWN0ZXItY2xhc3MiOjIsIm5vLWVtcHR5LXBhdHRlcm4iOjIsIm5vLWVtcHR5IjoyLCJuby1leC1hc3NpZ24iOjIsIm5vLWV4dHJhLWJvb2xlYW4tY2FzdCI6Miwibm8tZXh0cmEtc2VtaSI6Miwibm8tZmFsbHRocm91Z2giOjIsIm5vLWZ1bmMtYXNzaWduIjoyLCJuby1nbG9iYWwtYXNzaWduIjoyLCJuby1pbm5lci1kZWNsYXJhdGlvbnMiOjIsIm5vLWludmFsaWQtcmVnZXhwIjoyLCJuby1pcnJlZ3VsYXItd2hpdGVzcGFjZSI6Miwibm8tbWl4ZWQtc3BhY2VzLWFuZC10YWJzIjoyLCJuby1uZXctc3ltYm9sIjoyLCJuby1vYmotY2FsbHMiOjIsIm5vLW9jdGFsIjoyLCJuby1yZWRlY2xhcmUiOjIsIm5vLXJlZ2V4LXNwYWNlcyI6Miwibm8tc2VsZi1hc3NpZ24iOjIsIm5vLXNwYXJzZS1hcnJheXMiOjIsIm5vLXRoaXMtYmVmb3JlLXN1cGVyIjoyLCJuby11bmRlZiI6Miwibm8tdW5leHBlY3RlZC1tdWx0aWxpbmUiOjIsIm5vLXVucmVhY2hhYmxlIjoyLCJuby11bnNhZmUtZmluYWxseSI6Miwibm8tdW5zYWZlLW5lZ2F0aW9uIjoyLCJuby11bnVzZWQtbGFiZWxzIjoyLCJuby11bnVzZWQtdmFycyI6Miwibm8tdXNlbGVzcy1lc2NhcGUiOjIsInJlcXVpcmUteWllbGQiOjIsInVzZS1pc25hbiI6MiwidmFsaWQtdHlwZW9mIjoyLCJuby1yZXR1cm4tYXNzaWduIjoyfSwiZW52Ijp7fX19)

**What did you expect to happen?**

To have no lint error.

**What actually happened? Please include the actual, raw output from ESLint.**

It said: `Arrow function should not return assignment. (no-return-assign)`

**Comment by platinumazure** [comment]
Not sure that's doing what you might be thinking it does.

```js
[].map(result => (result['x'] = 'y', result))
```

This will take the result of `'y', result` (which is `result` after evaluating `'y'` for side effects) and assigning that to `result['x']`. In other words, it's the same as `result['x'] = result`.

And then, since the arrow function uses a concise body, that whole expression is returned.

If you're trying to assign to a member of `result` and then return the whole object, you need to do:

```js
[].map(result => { result['x'] = 'y'; return result; })
```

**Comment by EECOLOR** [comment]
@platinumazure I think you are mistaken, try it in a console:

```js
[{}].map(result => (result['x'] = 'y', result))
```
results in:
```js
[{ x: 'y' }]
```
---
Another example:
```js
[1, 2].reduce((result, x) => (result[x] = x, result), {})
```
results in:
```js
{
  1: '1',
  2: '2'
}
```

**Comment by platinumazure** [comment]
Never mind, I had wrong precedence. However, this is still a case of an assignment in a return statement (the return statement is implicit due to being a concise arrow body).

**Comment by not-an-aardvark** [comment]
I think the problem is that although the assignment is part of the returned `SequenceExpression`, the assignment itself is not the value getting returned.

**Comment by EECOLOR** [comment]
@platinumazure I get the feeling that nothing I will say will change your opinion.

The comma operator can be used for side effects (as I do in the examples) and explicitly states that the last statement is returned (or used as the value of the expression). If I wanted to disallow that I could enable the `no-sequences`.

If the assignment was in that last spot of the sequence I would expect the `no-return-assign` to trigger. In this case however the assignment is in a slot that was designed for side-effects.

**Comment by EECOLOR** [comment]
@platinumazure I'm sorry if I offended you, that was not my intention. It might just have been the text lacking human emotion.

In any case, what do you mean with the 'confused emoji`?

**Comment by platinumazure** [comment]
Nothing in particular. I just was surprised and confused that you had thought I was inflexible and unlikely to change my opinion given that I admitted to a mistaken understanding.

As a member of the team, it's my job to hold myself to high standards when interacting with the community. It seems I have failed to do so here. I apologize.

**Comment by EECOLOR** [comment]
@platinumazure I don't think there is a need to apologize. It might have been my prejudice (coming from other interactions about software) that mistakenly took the "However, this is still a case of an assignment in a return statement (the return statement is implicit due to being a concise arrow body)" as closing the door on the subject.

**Comment by EECOLOR** [comment]
Back to the subject. Do you think my argumentation is reasonable? The MDN documentation [explicitly](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Comma_Operator#Processing_and_then_returning) documents this use case and in my opinion there is a difference between these two forms: `return (a = b, c)` and `return x = y`.

**Comment by platinumazure** [comment]
I'm mildly opposed to changing the rule for this case, but I'm not going to stand in the way if others on the team believe this is a worthwhile change.

Would you mind editing your post to reflect [this rule change proposal template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-change-proposal.md)? And in exchange, I'll relabel this issue as a rule change proposal. Thanks!

**Comment by EECOLOR** [comment]
@platinumazure I changed the post.

**Comment by EECOLOR** [comment]
@platinumazure Did you have a chance to discuss this with the team?

**Comment by not-an-aardvark** [comment]
Thanks for your interest in improving ESLint. Unfortunately, it looks like this issue didn't get consensus from the team, so I'm closing it. We define [consensus](http://eslint.org/docs/maintainer-guide/issues.html#consensus) as having three :+1:s from team members, as well as a team member willing to champion the proposal. This is a high bar by design -- we can't realistically accept and maintain every feature request in the long term, so we only accept feature requests which are useful enough that there is consensus among the team that they're worth adding.

Since ESLint is pluggable and can load custom rules at runtime, the lack of consensus among the ESLint team doesn't need to be a blocker for you using this in your project, if you'd find it useful. It just means that you would need to [implement the rule](http://eslint.org/docs/developer-guide/working-with-rules) yourself, rather than using a bundled rule that is packaged with ESLint.

**Comment by juancampa** [comment]
Too bad this was closed. IMO this is a bug, it's not returning an assignment but the rule is detecting it as such.

One work-around is to use the `except-parens` setting and wrap your assignment in parens. It's rather ugly though
